chips/sam4l/src/usart.rs

```diff
@@ -1,131 +1,311 @@
+                                          pm::PBAClock::USART1,
+                                          nvic::NvicIdx::USART1,
+                                          dma::DMAPeripheral::USART1_RX,
+                                          dma::DMAPeripheral::USART1_TX);
+pub static mut USART2: USART = USART::new(USART_BASE_ADDRS[2],
+                                          pm::PBAClock::USART2,
+                                          nvic::NvicIdx::USART2,
+                                          dma::DMAPeripheral::USART2_RX,
+                                          dma::DMAPeripheral::USART2_TX);
+pub static mut USART3: USART = USART::new(USART_BASE_ADDRS[3],
+                                          pm::PBAClock::USART3,
+                                          nvic::NvicIdx::USART3,
+                                          dma::DMAPeripheral::USART3_RX,
+                                          dma::DMAPeripheral::USART3_TX);
 
 impl USART {
-    const fn new(location: Location, clock: PBAClock, nvic: nvic::NvicIdx) -> USART {
+    const fn new(base_addr: *mut USARTRegisters,
+                 clock: pm::PBAClock,
+                 nvic: nvic::NvicIdx,
+                 rx_dma_peripheral: dma::DMAPeripheral,
+                 tx_dma_peripheral: dma::DMAPeripheral)
+                 -> USART {
         USART {
-            regs: (BASE_ADDRESS + (location as usize) * SIZE) as *mut Registers,
-            clock: Clock::PBA(clock),
+            registers: base_addr,
+            clock: pm::Clock::PBA(clock),
             nvic: nvic,
-            dma: TakeCell::empty(),
-            dma_peripheral: DMAPeripheral::USART0_RX, // Set to some default.
-            // This is updated when a
-            // real DMA is configured.
-            client: None,
+
+            usart_rx_state: Cell::new(USARTStateRX::Idle),
+            usart_tx_state: Cell::new(USARTStateTX::Idle),
+
+            // these get defined later by `chip.rs`
+            rx_dma: TakeCell::empty(),
+            rx_dma_peripheral: rx_dma_peripheral,
+            rx_len: Cell::new(0),
+            tx_dma: TakeCell::empty(),
+            tx_dma_peripheral: tx_dma_peripheral,
+            tx_len: Cell::new(0),
+
+            // this gets defined later by `main.rs`
+            client: TakeCell::empty(),
         }
     }
 
-    pub fn set_client<C: uart::Client>(&mut self, client: &'static C) {
-        self.client = Some(client);
+    pub fn set_client(&self, client: &'static hil::uart::Client) {
+        self.client.replace(client);
     }
 
-    pub fn set_dma(&mut self, dma: &'static mut DMAChannel, dma_peripheral: DMAPeripheral) {
-        self.dma.replace(dma);
-        self.dma_peripheral = dma_peripheral;
+    pub fn set_dma(&self, rx_dma: &'static dma::DMAChannel, tx_dma: &'static dma::DMAChannel) {
+        self.rx_dma.replace(rx_dma);
+        self.tx_dma.replace(tx_dma);
     }
 
-    fn set_baud_rate(&self, baud_rate: u32) {
```

Comment:
It might be good to keep the `set_baud_rate()` function for future USART implementations (SPI, etc.).